### PR TITLE
Fix handle repr error with showlocals and verbose output

### LINKED ISSUE #5089: INTERNALERROR when error raised in __repr__ with showlocals and verbose output
The small test to reproduce

```python
class Thing:
    def __repr__(self):
        raise NotImplementedError

def test_the_thing():
    t = Thing()
    assert t == 3

def test_something_works():
    assert 1 == 1
```

When running tests with `$ python3 -m pytest` everything works as expected

```
================================== test session starts ===================================
platform linux -- Python 3.6.7, pytest-4.3.0, py-1.8.0, pluggy-0.9.0
rootdir: /home/kondratyev-nv/Development/repr_error, inifile:
plugins: describe-0.12.0
collected 2 items                                                                        

repr_error_test.py F.                                                          [100%]

======================================== FAILURES ========================================
_____________________________________ test_the_thing _____________________________________

    def test_the_thing():
        t = Thing()
>       assert t == 3
E       assert <[NotImplementedError("") raised in repr()] Thing object at 0x7fd26100aa58> == 3

repr_error_test.py:8: AssertionError
=========================== 1 failed, 1 passed in 0.03 seconds ===========================
```

Adding `-v` and `-l`  -  `$ python3 -m pytest` also works as expected

```
================================== test session starts ===================================
platform linux -- Python 3.6.7, pytest-4.3.0, py-1.8.0, pluggy-0.9.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/kondratyev-nv/Development/repr_error, inifile:
plugins: describe-0.12.0
collected 2 items                                                                        

repr_error_test.py::test_the_thing FAILED                                      [ 50%]
repr_error_test.py::test_something_works PASSED                                [100%]

======================================== FAILURES ========================================
_____________________________________ test_the_thing _____________________________________

    def test_the_thing():
        t = Thing()
>       assert t == 3
E       assert <[NotImplemen...x7fcd312745c0> == 3
E         (pytest_assertion plugin: representation of details failed.  Probably an object has a faulty __repr__.)
E         /home/kondratyev-nv/Development/repr_error/repr_error_test.py:3: NotImplementedError

t          = <[NotImplementedError("") raised in repr()] Thing object at 0x7fcd312745c0>

repr_error_test.py:8: AssertionError
=========================== 1 failed, 1 passed in 0.03 seconds ===========================
```

But increasing verbosity (`$ python3 -m pytest -l -vv`) leads to an INTERNALERROR.

```
================================== test session starts ===================================
platform linux -- Python 3.6.7, pytest-4.3.0, py-1.8.0, pluggy-0.9.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/kondratyev-nv/Development/repr_error, inifile:
plugins: describe-0.12.0
collected 2 items                                                                        

repr_error_test.py::test_the_thing 
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/main.py", line 210, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/main.py", line 250, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 68, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 62, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/main.py", line 271, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 68, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 62, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/runner.py", line 78, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/runner.py", line 93, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/runner.py", line 175, in call_and_report
INTERNALERROR>     report = hook.pytest_runtest_makereport(item=item, call=call)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 68, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/manager.py", line 62, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 203, in _multicall
INTERNALERROR>     gen.send(outcome)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/skipping.py", line 127, in pytest_runtest_makereport
INTERNALERROR>     rep = outcome.get_result()
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/runner.py", line 268, in pytest_runtest_makereport
INTERNALERROR>     longrepr = item.repr_failure(excinfo)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/python.py", line 837, in repr_failure
INTERNALERROR>     return self._repr_failure_py(excinfo, style=style)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/nodes.py", line 285, in _repr_failure_py
INTERNALERROR>     truncate_locals=truncate_locals,
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 556, in getrepr
INTERNALERROR>     return fmt.repr_excinfo(self)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 806, in repr_excinfo
INTERNALERROR>     reprtraceback = self.repr_traceback(excinfo)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 751, in repr_traceback
INTERNALERROR>     reprentry = self.repr_traceback_entry(entry, einfo)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 721, in repr_traceback_entry
INTERNALERROR>     localsrepr = self.repr_locals(entry.locals)
INTERNALERROR>   File "/home/kondratyev-nv/.local/lib/python3.6/site-packages/_pytest/_code/code.py", line 683, in repr_locals
INTERNALERROR>     str_repr = pprint.pformat(value)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 58, in pformat
INTERNALERROR>     compact=compact).pformat(object)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 144, in pformat
INTERNALERROR>     self._format(object, sio, 0, 0, {}, 0)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 161, in _format
INTERNALERROR>     rep = self._repr(object, context, level)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 393, in _repr
INTERNALERROR>     self._depth, level)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 405, in format
INTERNALERROR>     return _safe_repr(object, context, maxlevels, level)
INTERNALERROR>   File "/usr/lib/python3.6/pprint.py", line 555, in _safe_repr
INTERNALERROR>     rep = repr(object)
INTERNALERROR>   File "/home/kondratyev-nv/Development/repr_error/repr_error_test.py", line 3, in __repr__
INTERNALERROR>     raise NotImplementedError
INTERNALERROR> NotImplementedError

============================== no tests ran in 0.03 seconds ==============================
```

Also, notice that in the last example no tests are completed. I'm using Ubuntu 18.04, python 3.6.7, and pytest 4.3.0. 

Can you please clarify if this is expected behaviour or a bug?

### PR DESCRIPTION
**Pull request by kondratyev-nv**
Fixes #5089

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
nevermind, i missed a key part of the invocation site, its fine

File: src/_pytest/_io/saferepr.py [hunk 1]
```diff
@@ -1,8 +1,28 @@
...
 import sys
 
 from six.moves import reprlib
 
 
+def _safe_format_call(call, x, *args):
+    try:
+        # Try the vanilla repr and make sure that the result is a string
+        return call(x, *args)
+    except Exception:
```
Comment (RonnyPfannschmidt) [inline src/_pytest/_io/saferepr.py]: if the trackback is not used, just use `except Exception as e`

Comment (kondratyev-nv) [reply src/_pytest/_io/saferepr.py]: Just to make it clear, I have not changed this part, I've moved it from `_callhelper`. I thought `exc_info` was used to get the class of an exception. Would it be ok to use `e.__class__.__name__` instead?

Comment (kondratyev-nv) [reply src/_pytest/_io/saferepr.py]: @RonnyPfannschmidt Can you have a look, please? With `e.__class__.__name__` it looks like 

```
try:
    exc_name = e.__class__.__name__
except Exception:
    exc_name = "unknown"
try:
    exc_info = str(e)
except Exception:
    exc_info = "unknown"
```

File: src/_pytest/_code/code.py [hunk 2]
```diff
@@ -676,7 +679,7 @@ def repr_locals(self, locals):
                     if self.truncate_locals:
                         str_repr = self._saferepr(value)
                     else:
-                        str_repr = pprint.pformat(value)
+                        str_repr = self._safeformat(value)
```
Comment (blueyed) [inline src/_pytest/_code/code.py]: Why not use it directly?
I.e. not via `_safeformat` method - I see that this is done for `_saferepr`, but I do not think it is good to follow.

File: src/_pytest/_io/saferepr.py [hunk 3]
```diff
@@ -33,28 +53,20 @@ def repr_instance(self, x, level):
         return self._callhelper(repr, x)
 
     def _callhelper(self, call, x, *args):
-        try:
-            # Try the vanilla repr and make sure that the result is a string
-            s = call(x, *args)
-        except Exception:
-            cls, e, tb = sys.exc_info()
```
Comment (blueyed) [inline src/_pytest/_io/saferepr.py]: I'd keep the call here, and handle the exception via `_safe_format_exc`.

File: src/_pytest/_io/saferepr.py [hunk 4]
```diff
@@ -1,8 +1,28 @@
+import pprint
 import sys
 
 from six.moves import reprlib
 
 
+def _safe_format_call(call, x, *args):
```
Comment (blueyed) [inline src/_pytest/_io/saferepr.py]: This should be renamed to something more explicit like `_call_and_format_exception`.

File: src/_pytest/_io/saferepr.py (hunk 1 above)
Comment (blueyed) [reply src/_pytest/_io/saferepr.py]: Makes sense.  I'd use `as exc` though.

File: src/_pytest/_code/code.py (hunk 2 above)
Comment (kondratyev-nv) [reply src/_pytest/_code/code.py]: Yeah, I've just followed the way it has been called. I will change this to a direct call of `safeformat` and `saferepr`. Thank you.

File: src/_pytest/_io/saferepr.py (hunk 3 above)
Comment (kondratyev-nv) [reply src/_pytest/_io/saferepr.py]: Could you please explain it a bit? Do you mean something like
```
try:
    # Try the vanilla repr and make sure that the result is a string
    s = call(x, *args)
except Exception as exc:
    return _safe_format_exception(exc, x)
else:
    ...
```
and use `_safe_format_exception` here and in `_call_and_format_exception`?

Comment (blueyed) [reply src/_pytest/_io/saferepr.py]: Oh, sorry - forgot to remove that comment, after noticing that it is used twice here now (was confused by the name also).

File: src/_pytest/_io/saferepr.py [hunk 5]
```diff
@@ -1,8 +1,29 @@
...
 
+def _call_and_format_exception(call, x, *args):
+    try:
+        # Try the vanilla repr and make sure that the result is a string
+        return call(x, *args)
+    except Exception as exc:
+        try:
+            exc_name = exc.__class__.__name__
+        except Exception:
+            exc_name = "unknown"
```
Comment (blueyed) [inline src/_pytest/_io/saferepr.py]: This is not covered in tests (https://codecov.io/gh/pytest-dev/pytest/compare/9eac4733c56cd1bac8abb6639e19976710fee139...4024a121e446f348c5ea839e002f3362b8c4978a/diff#D1-13).
And I think it is not possible to get an exception via this, is it?  (/cc @asottile)

Comment (asottile) [reply src/_pytest/_io/saferepr.py]: sadly it is:

```python
class CryParty(object):
    @property
    def __class__(self):
        raise TypeError('boom!')
```

```pycon
>>> CryParty()
<__main__.CryParty object at 0x7f2e3c41b8d0>
>>> CryParty().__class__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "t.py", line 4, in __class__
    raise TypeError('boom!')
TypeError: boom!
>>> type(CryParty()).__name__
'CryParty'
```

File: testing/test_session.py [hunk 6]
```diff
@@ -134,6 +134,30 @@ def test_implicit_bad_repr1(self):
             != -1
         )
 
+    def test_broken_repr_with(self, testdir):
```
Comment (asottile) [inline testing/test_session.py]: with_showlocals_verbose perhaps?

File: src/_pytest/_io/saferepr.py (hunk 5 above)
Comment (kondratyev-nv) [reply src/_pytest/_io/saferepr.py]: Awesome! Thank you so much! I saw a decrease in the coverage, and I was looking without any luck for the whole evening to find a way to raise an exception with `__class__`. So I was even going to remove `try .. except` block.

**Comment by kondratyev-nv** [comment]
I have not found any way for `type(exc).__name__` to raise an exception, so I've removed the corresponding `try .. except` block in `_call_and_format_exception` in src/_pytest/_io/saferepr.py. If there is any possibility for this code to fail, I will be glad to return it and add more tests.

**Comment by blueyed** [comment]
Thank you!

**Pull request merged**